src_core/dynamics/models/elasticity_neo_hookean.rs

```diff
@@ -75,6 +76,48 @@ impl NeoHookeanElasticity {
         false
     }
 
+    // General elastic density function: https://dl.acm.org/doi/pdf/10.1145/3306346.3322949#section.5 (8) and (9)
+    // With degradation: https://dl.acm.org/doi/pdf/10.1145/3306346.3322949#subsection.3.2 (3)
+    // With hardening: https://www.math.ucla.edu/~cffjiang/research/mpmcourse/mpmcourse.pdf#subsection.6.5 (87)
+    #[allow(non_snake_case)]
+    pub fn elastic_energy_density(
+        &self,
+        particle_phase: Real,
+        elastic_hardening: Real,
+        deformation_gradient: &Matrix<Real>,
+    ) -> Real {
+        let F = deformation_gradient;
+
+        let hardened_mu = self.mu * elastic_hardening;
+        let hardened_lambda = self.lambda * elastic_hardening;
+
+        // aka. bulk modulus
+        let kappa = 2.0 / 3.0 * hardened_mu + hardened_lambda;
+
+        let g_c = Self::phase_coeff(particle_phase);
+
+        let J = F.determinant();
+
+        let alpha = -1. / DIM as Real;
+
+        let J_alpha = J.powf(alpha);
+
+        let Psi_mu =
+            |F: Matrix<Real>| hardened_mu / 2. * ((F.transpose() * F).trace() - DIM as Real);
+        let Psi_mu = Psi_mu(J_alpha * F);
```

Comment:
The closure makes the code more difficult to read, especially since it’s only called once and bound to a variable with the same name.